### ISSUE TITLE
Support multiple `eval` datasets

### ISSUE THREAD

**Issue by hanlint** [mosaicml/composer#41]
## 🚀 Feature Request

For fine-tuning tasks (e.g. GLUE) and also many vision experiments, need to support multiple eval datasets. The metrics needed could be different across different datasets. 

## [Optional] Implementation

* support for `eval_dataloaders` as a `List`
* during the `eval` loop, run through multiple dataloaders and log the metrics for each dataset
* to support different metrics, may need to either (1) store the metric with the datasets, or (2) have the model's `metric` function return different metrics depending on the dataset.

**Comment by ajaysaini725** [comment]
A few things to consider:
- The output of `model.validate(eval_data_batch) -> Tuple[Any, Any]` gets passed straight into each metrics returned by `model.metrics()` so the `validate()` function needs to be compatible with all `EvalDataset`s. Also, we should definitely keep `validate()` as a part of the model definition because it relies on the forward pass logic of the model. 
- Ideally we don't want to specify anything data-related in the model class because the model class definition should be an isolated unit that works across many different training runs. It's better to have each `EvalDataset` specify the metrics it needs. The model class definitions should still have a `metrics()` function that returns all metrics which can be computed for the model and then each `EvalDataset` can specify exactly which metrics are relevant for that dataset (this can be done in `DatasetHparams` and also should be stored in `DataloaderSpec` so that it works with the `Trainer` `__init__`. Having the relevant metrics for a model be returned by the `model.metrics()` function prevents an `EvalDataset` from specifying a metric that doesn't work for a particular model (i.e. the `model.metrics()` specifies all metrics for a model and then `EvalDataset` specifies the subset of metrics to use for that dataset). If `EvalDataset` doesn't specify anything then the default is to use all metrics. 


Also, another thought: what if we move the `train_dataset` and `eval_dataset` to be parameters of `Trainer.fit()` rather than parameters of `Trainer` instantiation?

**Comment by anisehsani** [comment]
Makes sense - I like the idea of making the datasets parameters to the fit function
